### PR TITLE
Indexed field access for List

### LINKED ISSUE #1005: Support nested list and map (dictionary) access
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
In structured formats, it's possible to have nested arrays and maps. In Spark for instance it's possible to do : ```select my_list[0][0]```

**Describe the solution you'd like**
Access deeply nested values

**Describe alternatives you've considered**
I saw there is a new Map DataType in arrow-rs 6.0, but I made this before so I'm not sure where to go from there.

**Additional context**
I have a nested array in an Avro record I use, which is a list of tuples, a tuple is a price and a quantity of something. Sometimes I only want to get the first value of that array because I don't care about the rest.

### PR DESCRIPTION
**Pull request by Igosuki**
# Which issue does this PR close?

Closes #1005 

 # Rationale for this change
Supporting accessing nested values is a great improvement as it makes everything more flexible since input data doesn't need to be flat, and so it doesn't need to be transformed by ETL prior to being ingested by datafusion.

# What changes are included in this PR?
Nested value access for Lists and Dictionary

# Are there any user-facing changes?
users will now be able to do : ```list[0][0]``` and ```dict['foo']['bar']```

No breaking changes.

I need to test the dictionary access I've only used the list one.

### CODE REVIEW

**Comment by Igosuki** [comment]
I had to patch sqlparser to make this work, in order for sqlparser to support numbers as field selector.

**Comment by houqp** [comment]
@Igosuki could you send your sqlparser-rs patch to upstream? We can help review release a new version to crates.io.

**Comment by Igosuki** [comment]
https://github.com/sqlparser-rs/sqlparser-rs/pull/356

**Comment by alamb** [comment]
I will try and get sql parser change merged in and a new release created something this weekend

**Comment by alamb** [comment]
This PR (that we never completed) from @jorgecarleitao  has related work I think -- namely access to struct fields: https://github.com/apache/arrow-datafusion/pull/628

**Comment by Igosuki** [comment]
I had initially used it, but it only allowed to use fields of a depth of
one, and was simply patching the compound field access.
I could look into allowing `table.col[i].field[j]` for instance in addition
to `table.col[i]['field'][j]'

Le ven. 17 sept. 2021 à 12:52, Andrew Lamb ***@***.***> a
écrit :

**Comment by Igosuki** [comment]
Will get back to this soon once I sort out the related issue here https://github.com/sqlparser-rs/sqlparser-rs/pull/356

**Comment by Igosuki** [comment]
I updated sqlparser for arbitrary nested access. But I'm having trouble understanding how select on dict should behave for instance with this https://github.com/apache/arrow-datafusion/pull/1006/commits/e2c00f145915c637e891132465a5e76b9762edeb#diff-6a157f1b320a619e7a3896027ce7d47a8e3ffeeae8ac3d21fe9157c92ebc2a6bL4955

**Comment by alamb** [comment]
I am preparing a sqlparser release FWIW: https://github.com/sqlparser-rs/sqlparser-rs/pull/360

**Comment by houqp** [comment]
@Igosuki do you need any help finishing up this PR?

**Comment by houqp** [comment]
I think there might be some confusions here. In my mind, `col["a"]["b"]` should access col as a dictionary in the sense of hashmap, not Arrow dictionary encoded array. In other words, when index key is a string, I think we should be matching for Struct datatype instead of dictionary type.

BTW, I am perfectly fine with us splitting the PRs into two and handle the dictionary access as a follow up if you prefer to do so :+1:

File: datafusion/src/physical_plan/expressions/get_indexed_field.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        }
+    }
+}
+
+/// Create a `.[field]` expression
+pub fn get_indexed_field(
+    arg: Arc<dyn PhysicalExpr>,
+    key: ScalarValue,
+) -> Result<Arc<dyn PhysicalExpr>> {
+    Ok(Arc::new(GetIndexedFieldExpr::new(arg, key)))
```
[the remark this answers is not in the document]
Comment (houqp) [reply datafusion/src/physical_plan/expressions/get_indexed_field.rs:110]: Good call, the result wrapping is not needed here.

**Comment by houqp** [comment]
Agree with @Jimexist 's comment, the rest looks good to me. @Igosuki for the follow up struct field access support PR, you can follow the physical plan evaluation implementation that @jorgecarleitao started awhile back at: https://github.com/apache/arrow-datafusion/pull/628/files#diff-ae63212535ac34ea66a26dc4726715b594b05794376b1b1fb511ff32211ed576R75. I think your SQL syntax combined with that physical plan implementation should be solid :+1:

**Comment by alamb** [comment]
Thanks for sticking with it @Igosuki !

**Pull request merged**